File: classification/exp_eval.m
function result = exp_eval()

% ---------------------------------------------------------
% This file is part of VISDA-17 challenge code. 
% This function is to calculate the mean accuarcy. 
% Please modify the following paths accordingly when you
% call this function:
% 1. pred_file_path : path to your prediction text file
% 2. gt_file_path   : path to the ground truth text file 
% Contact: Xingchao Peng (xpeng@bu.edu)
% ---------------------------------------------------------

% ---------------- User Configuration----------------------
pred_file_path = './result/example_prediction.txt';
gt_path = './result/val_ground_truth.txt';
% ------------------End of configuration--------------------


pred_label = textread(pred_file_path);
gt_label = textread(gt_path);
category_num  = 12;
category = {'aeroplane', 'bicycle', 'bus', 'car', 'horse', 'knife', ...
    'motorcycle' , 'person',  'plant', 'skateboard',  'train', 'truck'};


[Confmat,index] = confusionmat(pred_label, gt_label);
x_sum = sum(Confmat);
Confmat = double(Confmat);
sub_accuracy=zeros(category_num,1);

fprintf('Sub-accuracy:\n');
fprintf('=================\n');
for j =1:category_num
	Confmat(:,j) = Confmat(:,j)/x_sum(j);
	sub_accuracy(j) = Confmat(j,j);
	fprintf(['%10s | %0.1f\n'], category{j} ,sub_accuracy(j)*100);
end

fprintf('\n\n');
fprintf('Mean accuracy: %.2f\n', 100*sum(sub_accuracy)/length(sub_accuracy));

File: classification/exp_eval.py
#############################################################################
# This file is part of VISDA-17 challenge code for the classification track.
# It calculates the per-category and mean accuracy of your predictions 
# compared to the ground truth.
#
# Please modify the following paths accordingly when you
# call this function:
# 1. ground_truth: path to the ground truth text file
# 2. predictions: path to the text file with your predictions 
#
#############################################################################

from __future__ import division 
import numpy as np 
import sys, os, os.path


class categories():
    def __init__(self, names):
        self.names = names
        self.num_cat = len(names)
        self.acceptable_predictionss = []
        for i in range(len(self.names)):
            self.acceptable_predictionss.append(i)

        self.truth = np.zeros(self.num_cat)
        self.predictions = np.zeros(self.num_cat)
        self.predictions_accuracy = np.zeros(self.num_cat)

        for i in range(self.num_cat):
            self.truth[i] = 0
            self.predictions[i] = 0
            self.predictions_accuracy[i] = 0

        self.mean_predictions_accuracy = 0


###########################################################################
# classification_evaluation 
# 
# inputs: ground_truth.txt and predictions.txt files
#
# output: per-category and mean accuracies printed in 'scores.txt'
#
###########################################################################

def classification_evaluation(ground_truth,predictions):

    category_names = ['aeroplane', 'bicycle', 'bus', 'car', 'horse', 'knife', 'motorcycle', 'person', 'plant', 'skateboard', 'train', 'truck']
    categs = categories(category_names)

    with open(ground_truth) as f:
        truth = [x.strip('\n') for x in f.readlines()]
    f.close()

    with open(predictions) as f:
        predictions = [x.strip('\n') for x in f.readlines()]
    f.close()

    if len(predictions) == 0:
        print 'Error: predictions file is empty.'
        return

    if len(truth) != len(predictions):
        print 'Error: predictions file does not contain the same number of elements as truth file.'
        return

    for idx, category in enumerate(truth):
        truth_category = int(category)
        categs.truth[truth_category] += 1
        predictions_category = int(predictions[idx])

        if predictions_category not in categs.acceptable_predictionss:
            print 'Error: predictions file contains invalid entry. Please check that all category labels are valid and that the file adheres to the specified format for evaluation.'
            return

        if predictions_category == truth_category:
            categs.predictions[truth_category] += 1

    for i in range(categs.num_cat):
        if categs.truth[i] != 0:
            categs.predictions_accuracy[i] = 100*float(categs.predictions[i]/categs.truth[i])

    categs.mean_predictions_accuracy = float(np.mean(categs.predictions_accuracy))

    with open('scores.txt', 'wb') as f:
        f.write('mean accuracy: ' + str(categs.mean_predictions_accuracy) + '\n\n')
        for i in range(len(categs.predictions_accuracy)):
            f.write(categs.names[i] + ': ' + str(categs.predictions_accuracy[i]) + '\n\n')
    f.close()
    

if __name__ == '__main__':

    ground_truth = 'ground_truth.txt'
    predictions = 'predictions.txt'

    classification_evaluation(ground_truth,predictions)
